### PR TITLE
Add tinyint/smallint/integer/bigint to double coercion for hive tables

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Add `tinyint`/`smallint`/`integer`/`bigint` to `double` coercion for hive tables.

Enable the Hive users to retrieve the information from their tables after performing queries which change numeric integer columns to `double`.

Hive sample DDL query:
```
ALTER TABLE mytable CHANGE COLUMN mycolumn mycolumn double;
```

## Additional context and related issues

This is a mere adaptation of https://github.com/trinodb/trino/pull/16869 to match `timestamp` to `date` coercion

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
() Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Hive
* Support tinyint/smallint/integer/bigint to double schema evolution. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by dain** [comment]
Does the schema evolution match the behavior of Hive?

**Comment by dain** [comment]
To answer my own question, I think [this](https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/typeinfo/TypeInfoUtils.java#L949) is the method that Hive uses to determine if a type schema type is allowed.  The last lines of that code aren't really clear, but I believe it is says that anything on this list is coercible to anything higher on the list:

1. PrimitiveCategory.BYTE
2. PrimitiveCategory.SHORT
3. PrimitiveCategory.INT
4. PrimitiveCategory.LONG
5. PrimitiveCategory.DECIMAL
6. PrimitiveCategory.FLOAT
7. PrimitiveCategory.DOUBLE
8. PrimitiveCategory.STRING

**Comment by findinpath** [comment]
Unrelated CI failures
- `TestHiveTransactionalTable > testUnpartitionedDeleteAll(0: TRINO, 1: TRINO) [groups: hive_transactional]` https://github.com/trinodb/trino/issues/18216
- `TestIcebergSparkCompatibility > testRegisterTableWithComments(0: PARQUET) [groups: profile_specific_tests, iceberg]` 
```
io.trino.tempto.query.QueryExecutionException: java.sql.SQLException: Query failed (#20231024_225456_00532_2pkqk): More than one latest metadata file found at location: hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_comments_parquet_vi7gmeoqbt-93efd287b14c4ac7a78f42f7bf0f227c/metadata, latest metadata files are [hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_comments_parquet_vi7gmeoqbt-93efd287b14c4ac7a78f42f7bf0f227c/metadata/00006-a44edbac-5800-4500-9100-bf0323ad38a8.metadata.json, hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_comments_parquet_vi7gmeoqbt-93efd287b14c4ac7a78f42f7bf0f227c/metadata/00006-cb7496dc-adca-4712-be81-27677393b54b.metadata.json]
	at io.trino.tempto.query.JdbcQueryExecutor.execute(JdbcQueryExecutor.java:119)
```

**Comment by dain** [comment]
Is there an umbrella issue to track the missing coersions?

**Comment by findinpath** [comment]
> Is there an umbrella issue to track the missing coersions?

Created https://github.com/trinodb/trino/issues/19643 for providing of the past & ongoing schema evolution efforts in Hive.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java [hunk 1]
```diff
@@ -127,14 +127,32 @@ public static Type createTypeFromCoercer(TypeManager typeManager, HiveType fromH
...
+            if (toHiveType.equals(HIVE_LONG)) {
+                return Optional.of(new IntegerToBigintCoercer());
+            }
+            if (toHiveType.equals(HIVE_DOUBLE)) {
+                return Optional.of(new IntegerNumberToDoubleCoercer<>(fromType));
+            }
         }
-        if (fromHiveType.equals(HIVE_INT) && toHiveType.equals(HIVE_LONG)) {
-            return Optional.of(new IntegerToBigintCoercer());
+        if (fromHiveType.equals(HIVE_LONG) && (toHiveType.equals(HIVE_DOUBLE))) {
```
Comment (Praveen2112) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/coercions/CoercionUtils.java:154]: fromHiveType.equals(HIVE_LONG) && toHiveType.equals(HIVE_DOUBLE))
```suggestion
        if (fromHiveType.equals(HIVE_LONG) && toHiveType.equals(HIVE_DOUBLE)) {
```

**Review by Praveen2112 — APPROVED** [review]

**Review by dain — APPROVED** [review]

**Pull request merged**